okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,238 @@
+    private final ScheduledExecutorService schedulingExecutor;
+    private final ExecutorService executionExecutor;
+    private final Duration timeout;
+
+    private final int maxNumRelocations;
+
+    RemotingOkHttpCall(
+            Call delegate,
+            BackoffStrategy backoffStrategy,
+            UrlSelector urls, RemotingOkHttpClient client,
+            ScheduledExecutorService schedulingExecutor,
+            ExecutorService executionExecutor,
+            Duration timeout, int maxNumRelocations) {
+        super(delegate);
+        this.backoffStrategy = backoffStrategy;
+        this.urls = urls;
+        this.client = client;
+        this.schedulingExecutor = schedulingExecutor;
+        this.executionExecutor = executionExecutor;
+        this.timeout = timeout;
+        this.maxNumRelocations = maxNumRelocations;
+    }
+
+    private void scheduleExecution(Callback callback, Duration backoff) {
+        schedulingExecutor.schedule(
+                () -> executionExecutor.submit(() -> doClone().enqueue(callback)),
+                backoff.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
+
+    private void doEnqueueAndHandleErrorsAndRetries(Callback callback) {
+        super.enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException exception) {
+                if (!backoffStrategy.nextBackoff().isPresent()) {
+                    callback.onFailure(call, new IOException("Failed to complete the request due to an "
+                            + "IOException", exception));
+                } else {
+                    Optional<HttpUrl> redirectTo = urls.redirectToNext(request().url());
+                    if (!redirectTo.isPresent()) {
+                        callback.onFailure(call, new IOException("Failed to determine valid failover URL"
+                                + "for '" + request().url() + "' and base URLs " + urls.getBaseUrls()));
+                    } else {
+                        Request redirectedRequest = request().newBuilder()
+                                .url(redirectTo.get())
+                                .build();
+                        client.newCallWithMutableState(redirectedRequest, backoffStrategy, maxNumRelocations - 1)
+                                .enqueue(callback);
+                    }
+                }
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                // Relay successful responses
+                if (response.code() / 100 == 2) {
+                    callback.onResponse(call, response);
+                }
+
+                // Handle to handle QoS situations: retry, failover, etc.
+                Optional<QosException> qosError = qosHandler.handle(response);
+                if (qosError.isPresent()) {
+                    qosError.get().accept(new QosException.Visitor<Void>() {
```

Comment:
I think it would help readability to pull this chunk into a method, that way the `onResponse` will read very clearly as: success=>earlyreturn, hand qos, handle remoteexception, handle anything else,

Reply:
done